File: native-modules/react-native-range-downloader/android/src/main/java/com/margelo/nitro/reactnativerangedownloader/FirmwareArtifactReader.kt
package com.margelo.nitro.reactnativerangedownloader

import java.io.RandomAccessFile
import java.util.UUID

internal data class FirmwareArtifactReaderOpenResult(
  val readerId: String,
  val size: Long,
  val immutableToken: String,
)

internal class FirmwareArtifactReaderException private constructor(
  val code: String,
) : Exception(code) {
  companion object {
    fun notFinal() = FirmwareArtifactReaderException("ARTIFACT_NOT_FINAL")
    fun immutableTokenMismatch() =
      FirmwareArtifactReaderException("ARTIFACT_IMMUTABLE_TOKEN_MISMATCH")
    fun readerNotFound() = FirmwareArtifactReaderException("ARTIFACT_READER_NOT_FOUND")
    fun readerBusy() = FirmwareArtifactReaderException("ARTIFACT_READER_BUSY")
    fun invalidRange() = FirmwareArtifactReaderException("ARTIFACT_READ_INVALID_RANGE")
    fun readLimitExceeded() =
      FirmwareArtifactReaderException("ARTIFACT_READ_LIMIT_EXCEEDED")
    fun emptyRead() = FirmwareArtifactReaderException("ARTIFACT_READ_EMPTY")
    fun shortRead() = FirmwareArtifactReaderException("ARTIFACT_READ_SHORT")
    fun storageFailure() = FirmwareArtifactReaderException("ARTIFACT_STORAGE_FAILED")
  }
}

internal class FirmwareArtifactReader(
  private val store: FirmwareArtifactStore,
) : AutoCloseable {
  private class ReaderState(
    val readerId: String,
    val artifactRef: String,
    val immutableToken: String,
    val size: Long,
    val file: RandomAccessFile,
    val activityToken: FirmwareArtifactActivityToken,
  ) {
    var isReading = false
    var closeRequested = false

    fun close() {
      runCatching { file.close() }
      activityToken.close()
    }
  }

  private val lock = Any()
  private val readers = mutableMapOf<String, ReaderState>()

  fun open(
    artifactRef: String,
    immutableToken: String,
  ): FirmwareArtifactReaderOpenResult {
    val activityToken = store.beginActivity(
      artifactRef = artifactRef,
      kind = ArtifactActivityKind.READER,
    )
    try {
      val metadata = store.loadArtifact(activityToken.artifactRef)
      val actualSize = metadata.actualSize
      val actualSha256 = metadata.actualSha256
      val storedToken = metadata.immutableToken
      val isFinalClass = metadata.retentionClass in setOf(
        ArtifactRetentionClass.VERIFIED_FINAL,
        ArtifactRetentionClass.ARCHIVE,
        ArtifactRetentionClass.MATERIALIZED_CHILD,
      )
      if (metadata.tombstonedAtMillis != null ||
        !isFinalClass ||
        actualSize == null ||
        actualSize <= 0 ||
        actualSize != metadata.expectedSize ||
        actualSha256 == null ||
        !RangeDownloadLogic.secureHexEquals(actualSha256, metadata.expectedSha256) ||
        storedToken == null
      ) {
        throw FirmwareArtifactReaderException.notFinal()
      }
      if (immutableToken != storedToken) {
        throw FirmwareArtifactReaderException.immutableTokenMismatch()
      }

      val finalFile = store.finalFile(metadata.artifactRef)
      if (!finalFile.isFile || finalFile.length() != actualSize) {
        throw FirmwareArtifactReaderException.notFinal()
      }
      val randomAccessFile = RandomAccessFile(finalFile, "r")
      val readerId = UUID.randomUUID().toString().lowercase()
      val state = ReaderState(
        readerId = readerId,
        artifactRef = metadata.artifactRef,
        immutableToken = storedToken,
        size = actualSize,
        file = randomAccessFile,
        activityToken = activityToken,
      )
      synchronized(lock) {
        readers[readerId] = state
      }
      store.updateArtifact(metadata.artifactRef) { artifact ->
        artifact.lastAccessedAtMillis = System.currentTimeMillis()
      }
      return FirmwareArtifactReaderOpenResult(
        readerId = readerId,
        size = actualSize,
        immutableToken = storedToken,
      )
    } catch (error: Exception) {
      activityToken.close()
      throw mapOpenError(error)
    }
  }

  fun read(
    readerId: String,
    offset: Double,
    length: Double,
  ): ByteArray {
    val normalizedReaderId = validateReaderId(readerId)
    val validatedOffset = validateSafeInteger(offset)
    val validatedLength = validateSafeInteger(length)
    if (validatedLength > MAX_READ_BYTES) {
      throw FirmwareArtifactReaderException.readLimitExceeded()
    }

    val state = synchronized(lock) {
      val current = readers[normalizedReaderId]
        ?: throw FirmwareArtifactReaderException.readerNotFound()
      if (current.closeRequested) {
        throw FirmwareArtifactReaderException.readerNotFound()
      }
      if (current.isReading) {
        throw FirmwareArtifactReaderException.readerBusy()
      }
      if (validatedOffset > current.size ||
        validatedLength > current.size - validatedOffset
      ) {
        throw FirmwareArtifactReaderException.invalidRange()
      }
      if (validatedLength == 0L && validatedOffset != current.size) {
        throw FirmwareArtifactReaderException.emptyRead()
      }
      current.isReading = true
      current
    }

    try {
      if (validatedLength == 0L) {
        return ByteArray(0)
      }
      val bytes = ByteArray(validatedLength.toInt())
      state.file.seek(validatedOffset)
      val read = state.file.read(bytes)
      if (read != bytes.size) {
        throw FirmwareArtifactReaderException.shortRead()
      }
      return bytes
    } catch (error: FirmwareArtifactReaderException) {
      throw error
    } catch (error: Exception) {
      throw FirmwareArtifactReaderException.storageFailure()
    } finally {
      finishRead(state)
    }
  }

  fun close(readerId: String) {
    val normalizedReaderId = validateReaderId(readerId)
    val stateToClose = synchronized(lock) {
      val state = readers[normalizedReaderId] ?: return
      if (state.isReading) {
        state.closeRequested = true
        null
      } else {
        readers.remove(normalizedReaderId)
        state
      }
    }
    stateToClose?.close()
  }

  override fun close() {
    val statesToClose = synchronized(lock) {
      val idle = readers.values.filterNot { it.isReading }
      readers.values.filter { it.isReading }.forEach { it.closeRequested = true }
      idle.forEach { readers.remove(it.readerId) }
      idle
    }
    statesToClose.forEach { it.close() }
  }

  private fun finishRead(state: ReaderState) {
    val shouldClose = synchronized(lock) {
      state.isReading = false
      if (state.closeRequested) {
        readers.remove(state.readerId)
        true
      } else {
        false
      }
    }
    if (shouldClose) {
      state.close()
    }
  }

  private fun validateReaderId(readerId: String): String {
    val normalized = runCatching { UUID.fromString(readerId).toString().lowercase() }
      .getOrElse { throw FirmwareArtifactReaderException.readerNotFound() }
    if (!normalized.equals(readerId, ignoreCase = true)) {
      throw FirmwareArtifactReaderException.readerNotFound()
    }
    return normalized
  }

  private fun validateSafeInteger(value: Double): Long {
    if (!value.isFinite() ||
      value < 0 ||
      value > MAXIMUM_SAFE_INTEGER ||
      value % 1.0 != 0.0
    ) {
      throw FirmwareArtifactReaderException.invalidRange()
    }
    return value.toLong()
  }

  private fun mapOpenError(error: Exception): Exception =
    when (error) {
      is FirmwareArtifactReaderException -> error
      is FirmwareArtifactStoreException -> when (error.code) {
        "ARTIFACT_INVALID_REF", "ARTIFACT_NOT_FOUND" -> error
        else -> FirmwareArtifactReaderException.storageFailure()
      }
      else -> FirmwareArtifactReaderException.storageFailure()
    }

  companion object {
    const val MAX_READ_BYTES = 256L * 1024L
    private const val MAXIMUM_SAFE_INTEGER = 9_007_199_254_740_991.0

    @Volatile
    private var processReader: FirmwareArtifactReader? = null

    fun processInstance(store: FirmwareArtifactStore): FirmwareArtifactReader =
      processReader ?: synchronized(this) {
        processReader ?: FirmwareArtifactReader(store).also {
          processReader = it
        }
      }
  }
}

File: native-modules/react-native-range-downloader/ios/FirmwareArtifactReader.swift
import Foundation

struct FirmwareArtifactReaderOpenResult: Equatable {
  let readerId: String
  let size: Int64
  let immutableToken: String
}

enum FirmwareArtifactReaderError: Error, Equatable, LocalizedError {
  case notFinal
  case immutableTokenMismatch
  case readerNotFound
  case readerBusy
  case invalidRange
  case readLimitExceeded
  case emptyRead
  case shortRead
  case storageFailure

  var code: String {
    switch self {
    case .notFinal:
      return "ARTIFACT_NOT_FINAL"
    case .immutableTokenMismatch:
      return "ARTIFACT_IMMUTABLE_TOKEN_MISMATCH"
    case .readerNotFound:
      return "ARTIFACT_READER_NOT_FOUND"
    case .readerBusy:
      return "ARTIFACT_READER_BUSY"
    case .invalidRange:
      return "ARTIFACT_READ_INVALID_RANGE"
    case .readLimitExceeded:
      return "ARTIFACT_READ_LIMIT_EXCEEDED"
    case .emptyRead:
      return "ARTIFACT_READ_EMPTY"
    case .shortRead:
      return "ARTIFACT_READ_SHORT"
    case .storageFailure:
      return "ARTIFACT_STORAGE_FAILED"
    }
  }

  var errorDescription: String? {
    return code
  }
}

final class FirmwareArtifactReader {
  static let maxReadBytes = 256 * 1024
  static let shared = FirmwareArtifactReader(store: .shared)

  private static let maximumSafeInteger = 9_007_199_254_740_991.0

  private final class ReaderState {
    let readerId: String
    let artifactRef: String
    let immutableToken: String
    let size: Int64
    let fileHandle: FileHandle
    let activityToken: FirmwareArtifactActivityToken
    var isReading = false
    var closeRequested = false

    init(
      readerId: String,
      artifactRef: String,
      immutableToken: String,
      size: Int64,
      fileHandle: FileHandle,
      activityToken: FirmwareArtifactActivityToken
    ) {
      self.readerId = readerId
      self.artifactRef = artifactRef
      self.immutableToken = immutableToken
      self.size = size
      self.fileHandle = fileHandle
      self.activityToken = activityToken
    }

    func close() {
      try? fileHandle.close()
      activityToken.release()
    }
  }

  private let store: FirmwareArtifactStore
  private let lock = NSLock()
  private var readers: [String: ReaderState] = [:]

  init(store: FirmwareArtifactStore) {
    self.store = store
  }

  deinit {
    closeAll()
  }

  func open(
    artifactRef: String,
    immutableToken: String
  ) throws -> FirmwareArtifactReaderOpenResult {
    let activityToken = try store.beginActivity(
      artifactRef: artifactRef,
      kind: .reader
    )
    do {
      let metadata = try store.loadArtifact(activityToken.artifactRef)
      guard metadata.tombstonedAt == nil,
            metadata.retentionClass == .verifiedFinal ||
              metadata.retentionClass == .archive ||
              metadata.retentionClass == .materializedChild,
            let actualSize = metadata.actualSize,
            actualSize > 0,
            actualSize == metadata.expectedSize,
            let actualSha256 = metadata.actualSha256,
            RangeDownloadLogic.secureHexEquals(
              actualSha256,
              metadata.expectedSha256
            ),
            let storedToken = metadata.immutableToken else {
        throw FirmwareArtifactReaderError.notFinal
      }
      guard immutableToken == storedToken else {
        throw FirmwareArtifactReaderError.immutableTokenMismatch
      }

      let finalURL = try store.finalFileURL(for: metadata.artifactRef)
      let attributes = try FileManager.default.attributesOfItem(
        atPath: finalURL.path
      )
      guard let fileSize = attributes[.size] as? NSNumber,
            fileSize.int64Value == actualSize else {
        throw FirmwareArtifactReaderError.notFinal
      }
      let fileHandle = try FileHandle(forReadingFrom: finalURL)
      let readerId = UUID().uuidString.lowercased()
      let state = ReaderState(
        readerId: readerId,
        artifactRef: metadata.artifactRef,
        immutableToken: storedToken,
        size: actualSize,
        fileHandle: fileHandle,
        activityToken: activityToken
      )
      lock.withFirmwareLock {
        readers[readerId] = state
      }
      _ = try store.updateArtifact(metadata.artifactRef) { artifact in
        artifact.lastAccessedAt = Date().timeIntervalSince1970
      }
      return FirmwareArtifactReaderOpenResult(
        readerId: readerId,
        size: actualSize,
        immutableToken: storedToken
      )
    } catch {
      activityToken.release()
      throw mapError(error)
    }
  }

  func read(
    readerId: String,
    offset: Double,
    length: Double
  ) throws -> Data {
    let normalizedReaderId = try validateReaderId(readerId)
    let validatedOffset = try validateSafeInteger(offset)
    let validatedLength = try validateSafeInteger(length)
    guard validatedLength <= Int64(Self.maxReadBytes) else {
      throw FirmwareArtifactReaderError.readLimitExceeded
    }

    let state: ReaderState = try lock.withFirmwareLock {
      guard let state = readers[normalizedReaderId],
            !state.closeRequested else {
        throw FirmwareArtifactReaderError.readerNotFound
      }
      guard !state.isReading else {
        throw FirmwareArtifactReaderError.readerBusy
      }
      guard validatedOffset <= state.size,
            validatedLength <= state.size - validatedOffset else {
        throw FirmwareArtifactReaderError.invalidRange
      }
      if validatedLength == 0, validatedOffset != state.size {
        throw FirmwareArtifactReaderError.emptyRead
      }
      state.isReading = true
      return state
    }

    defer {
      finishRead(state)
    }

    if validatedLength == 0 {
      return Data()
    }

    do {
      try state.fileHandle.seek(toOffset: UInt64(validatedOffset))
      let data = try state.fileHandle.read(
        upToCount: Int(validatedLength)
      ) ?? Data()
      guard data.count == Int(validatedLength) else {
        throw FirmwareArtifactReaderError.shortRead
      }
      return data
    } catch let error as FirmwareArtifactReaderError {
      throw error
    } catch {
      throw FirmwareArtifactReaderError.storageFailure
    }
  }

  func close(readerId: String) throws {
    let normalizedReaderId = try validateReaderId(readerId)
    var stateToClose: ReaderState?
    lock.withFirmwareLock {
      guard let state = readers[normalizedReaderId] else {
        return
      }
      if state.isReading {
        state.closeRequested = true
      } else {
        readers.removeValue(forKey: normalizedReaderId)
        stateToClose = state
      }
    }
    stateToClose?.close()
  }

  func closeAll() {
    var statesToClose: [ReaderState] = []
    lock.withFirmwareLock {
      for state in readers.values {
        if state.isReading {
          state.closeRequested = true
        } else {
          statesToClose.append(state)
        }
      }
      for state in statesToClose {
        readers.removeValue(forKey: state.readerId)
      }
    }
    statesToClose.forEach { $0.close() }
  }

  private func finishRead(_ state: ReaderState) {
    var shouldClose = false
    lock.withFirmwareLock {
      state.isReading = false
      if state.closeRequested {
        readers.removeValue(forKey: state.readerId)
        shouldClose = true
      }
    }
    if shouldClose {
      state.close()
    }
  }

  private func validateReaderId(_ readerId: String) throws -> String {
    guard let uuid = UUID(uuidString: readerId),
          uuid.uuidString.lowercased() == readerId.lowercased() else {
      throw FirmwareArtifactReaderError.readerNotFound
    }
    return uuid.uuidString.lowercased()
  }

  private func validateSafeInteger(_ value: Double) throws -> Int64 {
    guard value.isFinite,
          value >= 0,
          value <= Self.maximumSafeInteger,
          value.rounded(.towardZero) == value else {
      throw FirmwareArtifactReaderError.invalidRange
    }
    return Int64(value)
  }

  private func mapError(_ error: Error) -> Error {
    if let readerError = error as? FirmwareArtifactReaderError {
      return readerError
    }
    if let storeError = error as? FirmwareArtifactStoreError {
      switch storeError {
      case .invalidReference, .artifactNotFound:
        return storeError
      default:
        return FirmwareArtifactReaderError.storageFailure
      }
    }
    return FirmwareArtifactReaderError.storageFailure
  }
}
